### PR TITLE
Add protected-namespace-regex flag

### LINKED ISSUE #72: Support for protected-namespace-regexp option
# Describe the feature

We had a concept of protected namespaces, e.g. we do not allow users to create namespaces ending on -system$ , as they had special monitoring and other rules, which are applied to such NS. So I would like to add additional flag `--protected-namespace-regexp`  and modify tenant_prefix  webhook, so will also reject creation of namespaces, which are covered by this regexp.

### PR DESCRIPTION
**Pull request by MaxFedotov**
This PR closes #72 and adds configuration parameter `protected-namespace-regex` which disallow users to create namespaces whose name matches this regexp.

It's completely backward-compatible, as default `protected-namespace-regex` is set to empty string.

### CODE REVIEW

**Comment by prometherion** [comment]
Thanks for this feature proposal, looks promising!

At this stage, I would say is welcome, although I was thinking with @bsctl on how to make these kind of checks highly customisable using OPA/Gatekeeper directly from the tenant or Capsule configuration: I would start writing down this enhancement proposal in order to track it.

I'll take a look to this PR ASAP, in the meanwhile: many kudos!

**Comment by MaxFedotov** [comment]
@prometherion 
thanks a lot :)

**Comment by MaxFedotov** [comment]
strange thing, when I ran tests locally everything is OK
```
Ran 18 of 18 Specs in 101.191 seconds
SUCCESS! -- 18 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
```
But It failes on github actions

**Comment by prometherion** [comment]
> seems like it is timeout issue due to pod being recreated when I modify deployment args. Will increase defaultTimeoutInterval and see if it will be OK

Honestly, I would increase just the timeout interval for the Pod recreation, I'd like to keep tests as fast as possible, if you don't mind.

File: e2e/utils_test.go [hunk 1]
```diff
@@ -23,15 +23,17 @@ import (
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/clastix/capsule/api/v1alpha1"
 )
 
 const (
-	defaultTimeoutInterval = 10 * time.Second
+	defaultTimeoutInterval = 35 * time.Second
```
Comment (prometherion) [inline e2e/utils_test.go:36]: Please, don't increase this but rather set an override for the Pod recreation one.

**Review by prometherion — CHANGES_REQUESTED** [review]
Except the increased timeout for tests, LGTM: what do you think @bsctl? I guess this feature is really promising!

**Comment by bsctl** [comment]
Just provide an example for documenting the feature. Thanks

**Comment by MaxFedotov** [comment]
> Just provide an example for documenting the feature. Thanks

I've already added it to README.md
```
You can disallow users to create namespaces matching a particular regexp by passing `--protected-namespace-regex` option with a value of regular expression.
```

File: e2e/utils_test.go (hunk 1 above)
Comment (MaxFedotov) [reply e2e/utils_test.go]: fixed, moved it to different variable `podRecreationTimeoutInterval`

**Comment by MaxFedotov** [comment]
@prometherion 
rebased on master, local tests OK
```
Running Suite: Controller Suite
===============================
Random Seed: 1598960277
Will run 20 of 20 specs

STEP: bootstrapping test environment
2020-09-01T14:38:01.044+0300    DEBUG   controller-runtime.test-env     using existing cluster
2020-09-01T14:38:01.044+0300    DEBUG   controller-runtime.test-env     automatically acquiring client configuration
2020-09-01T14:38:01.045+0300    DEBUG   controller-runtime.test-env     installing CRDs
2020-09-01T14:38:01.045+0300    DEBUG   controller-runtime.test-env     reading CRDs from path  {"path": "../config/crd/bases"}
2020-09-01T14:38:01.051+0300    DEBUG   controller-runtime.test-env     read CRDs from file     {"file": "capsule.clastix.io_tenants.yaml"}
2020-09-01T14:38:01.355+0300    DEBUG   controller-runtime.test-env     installing CRD  {"crd": "tenants.capsule.clastix.io"}
2020-09-01T14:38:01.365+0300    DEBUG   controller-runtime.test-env     CRD already exists, updating    {"crd": "tenants.capsule.clastix.io"}
2020-09-01T14:38:01.394+0300    DEBUG   controller-runtime.test-env     adding API in waitlist  {"GV": "capsule.clastix.io/v1alpha1"}
2020-09-01T14:38:01.394+0300    DEBUG   controller-runtime.test-env     adding API in waitlist  {"GV": "capsule.clastix.io/v1alpha1"}
2020-09-01T14:38:01.396+0300    DEBUG   controller-runtime.test-env     installing webhooks
creating a Namespace over-quota 
  should fail
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/overquota_namespace_test.go:53
STEP: creating three Namespaces
•
------------------------------
creating a Namespace with --protected-namespace-regex enabled 
  should fail
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/protected_namespace_regex_test.go:51

• [SLOW TEST:28.357 seconds]
creating a Namespace with --protected-namespace-regex enabled
/Users/m_fedotov/Work/Code/k8s/capsule/e2e/protected_namespace_regex_test.go:29
  should fail
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/protected_namespace_regex_test.go:51
------------------------------
creating a Namespace with --protected-namespace-regex enabled 
  should succeed and be available in Tenant namespaces list
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/protected_namespace_regex_test.go:57

• [SLOW TEST:20.211 seconds]
creating a Namespace with --protected-namespace-regex enabled
/Users/m_fedotov/Work/Code/k8s/capsule/e2e/protected_namespace_regex_test.go:29
  should succeed and be available in Tenant namespaces list
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/protected_namespace_regex_test.go:57
------------------------------
when Tenant handles Ingress classes 
  should block non allowed Ingress class
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/ingress_class_test.go:64
STEP: non-specifying the class
STEP: using a forbidden class as Annotation
STEP: specifying a forbidden class
•
------------------------------
when Tenant handles Ingress classes 
  should allow enabled Ingress class using the deprecated Annotation
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/ingress_class_test.go:127
•
------------------------------
when Tenant handles Ingress classes 
  should allow enabled Ingress class using the IngressClassName field
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/ingress_class_test.go:155

• [SLOW TEST:6.246 seconds]
when Tenant handles Ingress classes
/Users/m_fedotov/Work/Code/k8s/capsule/e2e/ingress_class_test.go:38
  should allow enabled Ingress class using the IngressClassName field
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/ingress_class_test.go:155
------------------------------
when Tenant handles Storage classes 
  should block non allowed Storage Class
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/storage_class_test.go:61
STEP: non-specifying the class
STEP: specifying a forbidden class
•
------------------------------
when Tenant handles Storage classes 
  should allow enabled Storage Class
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/storage_class_test.go:107
•
------------------------------
changing Tenant managed Kubernetes resources 
  should reapply the original resources upon third party change
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/tenant_resources_changes_test.go:179
STEP: creating the Namespaces
STEP: changing Limit Range resources
STEP: changing Network Policy resources
STEP: changing Resource Quota resources
STEP: changing Limit Range resources
STEP: changing Network Policy resources
STEP: changing Resource Quota resources
STEP: changing Limit Range resources
STEP: changing Network Policy resources
STEP: changing Resource Quota resources
STEP: changing Limit Range resources
STEP: changing Network Policy resources
STEP: changing Resource Quota resources

• [SLOW TEST:17.898 seconds]
changing Tenant managed Kubernetes resources
/Users/m_fedotov/Work/Code/k8s/capsule/e2e/tenant_resources_changes_test.go:37
  should reapply the original resources upon third party change
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/tenant_resources_changes_test.go:179
------------------------------
when Tenant owner interacts with the webhooks 
  should disallow deletions
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/owner_webhooks_test.go:103
STEP: blocking Capsule Limit ranges
STEP: blocking Capsule Network Policy
STEP: blocking blocking Capsule Resource Quota
•
------------------------------
when Tenant owner interacts with the webhooks 
  should allow listing
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/owner_webhooks_test.go:147
STEP: Limit Range resources
STEP: Network Policy resources
STEP: Resource Quota resources
•
------------------------------
when Tenant owner interacts with the webhooks 
  should allow all actions to Tenant owner Network Policy resources
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/owner_webhooks_test.go:182
STEP: creating
STEP: updating
STEP: deleting
•
------------------------------
creating a Namespace as Tenant owner 
  should be available in Tenant namespaces list
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/new_namespace_test.go:53
•
------------------------------
creating namespaces within a Tenant with resources 
  should contains all replicated resources
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/tenant_resources_test.go:181
STEP: creating the Namespaces
STEP: checking Limit Range resources
STEP: checking Network Policy resources
STEP: checking the Namespace scheduler annotation
STEP: checking the Resource Quota resources
STEP: checking Limit Range resources
STEP: checking Network Policy resources
STEP: checking the Namespace scheduler annotation
STEP: checking the Resource Quota resources
STEP: checking Limit Range resources
STEP: checking Network Policy resources
STEP: checking the Namespace scheduler annotation
STEP: checking the Resource Quota resources
•
------------------------------
exceeding Tenant resource quota 
  should block new Pods if limit is reached
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/resource_quota_exceeded_test.go:136
STEP: creating the Namespaces
STEP: retrieving the Resource Quota
STEP: ensuring the status has been blocked with actual usage
STEP: creating an exceeded Pod
STEP: retrieving the Resource Quota
STEP: ensuring the status has been blocked with actual usage
STEP: creating an exceeded Pod

• [SLOW TEST:6.336 seconds]
exceeding Tenant resource quota
/Users/m_fedotov/Work/Code/k8s/capsule/e2e/resource_quota_exceeded_test.go:39
  should block new Pods if limit is reached
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/resource_quota_exceeded_test.go:136
------------------------------
creating a Namespace trying to select a third Tenant 
  should fail
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/selecting_non_owned_tenant_test.go:53
STEP: assigning to the Namespace the Capsule Tenant label
•
------------------------------
Namespace creation with no Tenant assigned 
  should fail
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/missing_tenant_test.go:32
•
------------------------------
creating a Namespace with Tenant selector 
  should be assigned to the selected Tenant
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/selecting_tenant_test.go:69
STEP: assigning to the Namespace the Capsule Tenant label
•
------------------------------
creating a Namespace as Tenant owner with custom --capsule-group 
  should fail
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/custom_capsule_group_test.go:54

• [SLOW TEST:22.185 seconds]
creating a Namespace as Tenant owner with custom --capsule-group
/Users/m_fedotov/Work/Code/k8s/capsule/e2e/custom_capsule_group_test.go:32
  should fail
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/custom_capsule_group_test.go:54
------------------------------
creating a Namespace as Tenant owner with custom --capsule-group 
  should succeed and be available in Tenant namespaces list
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/custom_capsule_group_test.go:61

• [SLOW TEST:19.223 seconds]
creating a Namespace as Tenant owner with custom --capsule-group
/Users/m_fedotov/Work/Code/k8s/capsule/e2e/custom_capsule_group_test.go:32
  should succeed and be available in Tenant namespaces list
  /Users/m_fedotov/Work/Code/k8s/capsule/e2e/custom_capsule_group_test.go:61
------------------------------
STEP: tearing down the test environment


Ran 20 of 20 Specs in 145.021 seconds
SUCCESS! -- 20 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
```

File: pkg/webhook/tenant_prefix/patching.go [hunk 2]
```diff
@@ -49,18 +51,26 @@ func (o Webhook) GetPath() string {
...
-		return admission.Allowed("")
-	}
 	ns := &corev1.Namespace{}
 	if err := decoder.Decode(req, ns); err != nil {
 		return admission.Errored(http.StatusBadRequest, err)
 	}
 
+	if r.protectedNamespacesRegex != nil {
+		if matched := r.protectedNamespacesRegex.MatchString(ns.GetName()); matched {
+			return admission.Denied("Creating namespaces with name matching " + r.protectedNamespacesRegex.String() + " regexp is not alloweded; please, reach out the system administrators")
```
Comment (prometherion) [inline pkg/webhook/tenant_prefix/patching.go:66]: Should the error be `Creating namespace with name non-matching `r.protectedNamespacesRegex.String()` regex is not allowed`?

Take a look also to the typo on `alloweded`

**Review by prometherion — CHANGES_REQUESTED** [review]
Except the typo and the error message, LGTM: maybe just rebase over `master` to run e2e tests according to latest fixes.

File: pkg/webhook/tenant_prefix/patching.go (hunk 2 above)
Comment (MaxFedotov) [reply pkg/webhook/tenant_prefix/patching.go]: fixed

**Review by prometherion — APPROVED** [review]
LGTM, astonishing feature @MaxFedotov: thanks a lot!

**Pull request merged**